### PR TITLE
[SPARK-36324][CORE] Replace revertPartialWritesAndClose with close in ExternalSorter.spill and ExternalAppendOnlyMap.spill

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
`ExternalAppendOnlyMap.spill` and `ExternalSorter.spill` call  `revertPartialWritesAndClose` method when `objectsWritten == 0` as follows:

``` scala
try {
      while (inMemoryIterator.hasNext) {
        ...

        if (objectsWritten == serializerBatchSize) {
          flush()
        }
      }
      if (objectsWritten > 0) {
        flush()
        writer.close()
      } else {
        writer.revertPartialWritesAndClose()
      }
      success = true
    } finally {
      ...
    }
```

When `objectsWritten` is 0,  all data has been `flush()` and nothing needs to be revert,  and `revertPartialWritesAndClose()`  will add some unnecessary file operations like file reopen and truncate.

### Why are the changes needed?
Replace `revertPartialWritesAndClose()` with `close()` in `ExternalSorter.spill` and `ExternalAppendOnlyMap.spill` to reduce some unnecessary file operations like file reopen and truncate.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. Pass the Jenkins or GitHub Action
2. Add new Test cases to verify `spill behavior not change`

### CODE REVIEW

**Comment by LuciferYang** [comment]
6f4430c and 6573bca add 2 case to test `ExternalAppendOnlyMap.spill` and `ExternalSorter.spill` methods, the test case verify `recordsWritten same as data size` and `bytesWritten same as file size`.

Then next commits will replace `revertPartialWritesAndClose` with `close.

**Comment by LuciferYang** [comment]
image

image

115345d Pass the Jenkins or GitHub Action

**Comment by LuciferYang** [comment]
cc @mridulm @Ngone51 @HyukjinKwon @dongjoon-hyun

**Comment by mridulm** [comment]
The assumption being made here is that if no objects are written, then nothing was written to the file.
This is not a general assumption to make (there could be some header written for example - not just right now, but in future as we add support for other serializers/codecs/etc).

I am assuming this is an effort at cleanup ? If yes, let us avoid making this change.
If there is some functional reason to make this (bug/severe perf issue/etc), please do let me know.

**Comment by LuciferYang** [comment]
> The assumption being made here is that if no objects are written, then nothing was written to the file.
This is not a general assumption to make (there could be some header written for example - not just right now, but in future as we add support for other serializers/codecs/etc).

https://github.com/apache/spark/blob/0ece865ea4b78f8144defcadd143fccf3dc99743/core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala#L291-L320

I mean, if the code runs to line 318,  all data has been `flush()`ed on line 311, so `revert Partial Writes` here does not change any structure of the file because the `reportedPosition = committedPosition` and `numRecordsWritten = 0` after `flush()`

On the other hand, if we add some meta to the file in the future, should the empty file also need contain these meta? If yes, should we also manual `flush()` meta first before `revertPartialWritesAndClose()`？

If we call `revertPartialWritesAndClose()` directly without manually `flush()` the meta first, the meta of the empty file length will also be truncate to 0 because only `flush()` change the `committedPosition`. And if we manually `flush()` the first, the `revertPartialWritesAndClose()` will no longer change the file structure and `close()` is enough.

**Comment by mridulm** [comment]
In general, do not couple implementation details/expectations to general collections.
If open() or flush() or close() can result in writes to the file, expectation of revert is to ensure this is discarded; if the code above evolves, we dont want to keep going down the stack to ensure correctness.

Unless I am seeing a logical or performance issue to make this change, I would not make it.

**Review by Ngone51 — COMMENTED** [review]
Could we just avoid file operations when `initialized=false` in `revertPartialWritesAndClose`?

**Comment by LuciferYang** [comment]
> Could we just avoid file operations when `initialized=false` in revertPartialWritesAndClose?

It seems that `initialized` is always `false` after `closeResources()`

**Comment by LuciferYang** [comment]
@Ngone51 However, I feel that these file operations seems redundant in some scenarios, except for the current pr

For example:

```
if (!success) {
        // This code path only happens if an exception was thrown above before we set success;
        // close our stuff and let the exception be thrown further
        writer.revertPartialWritesAndClose()
        if (file.exists()) {
          if (!file.delete()) {
            logWarning(s"Error deleting ${file}")
          }
        }
      }
```
I don't understand why we need to truncate a file before deleting an unnecessary file. Therefore, I also want to add a boolean arg to the `revertPartialWritesAndClose ` method to control whether to do truncate operation.

File: core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterCloseMethodBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      flush()
+      writer.close()
+    } else if (revertAndClose) {
+      writer.revertPartialWritesAndClose()
+    } else {
+      writer.close()
+    }
+  }
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterCloseMethodBenchmark.scala:80]: It is difficult to compare benchmark directly between `writer.close()` and `writer.revertPartialWritesAndClose()`, so I extracted the main code part of `ExternalSorter.spill` to compare the performance of `writer.close()` and `writer.revertPartialWritesAndClose()`

@mridulm @Ngone51

File: core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterCloseMethodBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    val conf = SparkEnv.get.conf
+    val bufferSize = conf.get(config.SHUFFLE_FILE_BUFFER_SIZE).toInt * 1024
+    val batchSize = conf.get(config.SHUFFLE_SPILL_BATCH_SIZE)
+    val numIters = 3
+
+
+    runBenchmark("emptyDataBuffer") {
+      val collection = new PartitionedPairBuffer[Int, Int]
+      val iter = collection.destructiveSortedWritablePartitionedIterator(None)
+      val benchmark = new Benchmark("emptyDataBuffer", 1, output = output)
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterCloseMethodBenchmark.scala:92]: The `close` or `revertPartialWritesAndClose` method is executed only once per iteration, so `valuesPerIteration` is set to 1

File: core/benchmarks/DiskBlockObjectWriterCloseMethodBenchmark-results.txt [hunk 3]
```diff
@@ -0,0 +1,48 @@
+================================================================================================
+emptyDataBuffer
```
Comment (LuciferYang) [inline core/benchmarks/DiskBlockObjectWriterCloseMethodBenchmark-results.txt:2]: Although the else branch is entered only when 'objectsWritten = = 0', there still seem to be some performance differences between the `close` and `revertPartialWritesAndClose` method.

I run the bench with ssd，I don't know if there will be a more obvious difference if HDD is used

File: core/benchmarks/DiskBlockObjectWriterCloseMethodBenchmark-results.txt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+================================================================================================
+
+OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Mac OS X 11.4
+Apple M1
+emptyDataBuffer:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+use close                                             0              0           0          0.1       11416.0       1.0X
+use revertPartialWritesAndClose                       0              0           0          0.0       50000.0       0.2X
```
Comment (LuciferYang) [inline core/benchmarks/DiskBlockObjectWriterCloseMethodBenchmark-results.txt:10]: Although the else branch is entered only when 'objectsWritten = = 0', there still seem to be some performance differences between the close and revertPartialWritesAndClose method.

I run the bench with ssd，I don't know if there will be a more obvious difference if HDD is used

**Comment by LuciferYang** [comment]
> In general, do not couple implementation details/expectations to general collections.
> If open() or flush() or close() can result in writes to the file, expectation of revert is to ensure this is discarded; if the code above evolves, we dont want to keep going down the stack to ensure correctness.
> 
> Unless I am seeing a logical or performance issue to make this change, I would not make it.

Can you give me more advice so that I can add a case for Spark that can only use the `revertPartialWritesAndClose` when  `objectsWritten = = 0,` which is also a harvest for me, thx @mridulm

**Comment by Ngone51** [comment]
>> Could we just avoid file operations when initialized=false in revertPartialWritesAndClose?
>
> It seems that initialized is always false after closeResources()

How can we call `closeResources` if `initialized` is already `false`?

**Comment by Ngone51** [comment]
> I don't understand why we need to truncate a file before deleting an unnecessary file. Therefore, I also want to add a boolean arg to the revertPartialWritesAndClose method to control whether to do truncate operation.

This's a known issue. You're right. It doesn't make sense to truncate the file since we'll delete the file anyway. It's probably due to the historical reason. But I don't like the idea to add a parameter to control the truncate operation. I don't think there's a valid case that we want "truncate" indeed. Do we? I think this's actually a separate issue beyond what this PR was trying to fix at the beginning. But I'm OK if you'd like to refactor the `revertPartialWritesAndClose` and fix both issues together.

**Comment by LuciferYang** [comment]
> This's a known issue. You're right. It doesn't make sense to truncate the file since we'll delete the file anyway. It's probably due to the historical reason. But I don't like the idea to add a parameter to control the truncate operation. I don't think there's a valid case that we want "truncate" indeed. Do we? I think this's actually a separate issue beyond what this PR was trying to fix at the beginning. But I'm OK if you'd like to refactor the revertPartialWritesAndClose and fix both issues together.

Thx for your explanation @Ngone51, I will try to fix it in another pr, this is independent of the current pr

**Comment by LuciferYang** [comment]
> Could we just avoid file operations when initialized=false in revertPartialWritesAndClose?

Or Could we just avoid file operations when `reportedPosition - committedPosition == 0` in revertPartialWritesAndClose?

**Comment by mridulm** [comment]
@Ngone51, @LuciferYang There are two cases here:
* No writes after flush
* No writes at all

(Both are almost the same - just calling them out).

In these, it is possible for (meta-) data to be written out during `close` even if `objectsWritten == 0`.
A `close` will implicitly `flush()` pending bytes and then `close` the output stream - potentially resulting in non-zero bytes being written out. (For an example, see `DeflaterOutputStream.finish`).

Currently, we do not have data in file which is not part of some segment - with the PR, we can have these suffix bytes in file.

Do let me know if I am missing something here.

**Comment by Ngone51** [comment]
> In these, it is possible for (meta-) data to be written out during close even if objectsWritten == 0.

@mridulm  In that case, shouldn't we add another flag to indicate if there's any metadata written? `objectsWritten` definitely wouldn't work.

**Comment by Ngone51** [comment]
>I mean closeResources() will set initialized as false and initialized is always false in finally block, so will "just avoid file >operations when initialized=false in revertPartialWritesAndClose" become "always not do these file operations" ?
>
>Utils.tryWithSafeFinally {
      if (initialized) {
        writeMetrics.decBytesWritten(reportedPosition - committedPosition)
        writeMetrics.decRecordsWritten(numRecordsWritten)
        streamOpen = false
        closeResources() // `closeResources()` will set `initialized` as false
      }
    } {
      //  `initialized` is always false in this block
      // do truncate file operations. 
    }

You're right...I was thinking about the case where there's no data at all..

**Comment by LuciferYang** [comment]
@mridulm thx for your explain

I test `DeflaterOutputStream`, there will be 8 bytes([120,-100,3,0,0,0,0,1]) here even if we don't write anything and I think I basically understand what you mean. 

However, I have another question：If this scenario is what we need to consider,  should we need call `revertPartialWritesAndClose ` when `  objectsWritten > 0` also?  otherwise these 8 bytes will also be written to the file ...

**Comment by LuciferYang** [comment]
If we consider that `close()` or other methods will write some suffix data to the file end, 

I try to write a case:

```scala
test("Close will wirte some meta data to os") {

    when(blockManager.getDiskWriter(
      any(classOf[BlockId]),
      any(classOf[File]),
      any(classOf[SerializerInstance]),
      anyInt(),
      any(classOf[ShuffleWriteMetrics])
    )).thenAnswer((invocation: InvocationOnMock) => {
      val args = invocation.getArguments
      val shuffleWriteMetrics = args(4).asInstanceOf[ShuffleWriteMetrics]
      metricsCreated += shuffleWriteMetrics
      new DiskBlockObjectWriter(
        args(1).asInstanceOf[File],
        blockManager.serializerManager,
        args(2).asInstanceOf[SerializerInstance],
        args(3).asInstanceOf[Int],
        false,
        shuffleWriteMetrics,
        args(0).asInstanceOf[BlockId]
      ) {
       // Suppose we can override this method, for example,  change private to protected
        override def closeResources(): Unit = {
          val meta = "suffix".getBytes("UTF-8") // write `suffix` to the file before close.
          new FileOutputStream(file, true).write(meta)
          super.closeResources()
        }
      }
    })

    // Test data size corresponds to three different scenarios:
    // 1. spillBatchSize -> `objectsWritten == 0`
    // 2. spillBatchSize + 1 -> `objectsWritten > 0`
    // 3. 0 -> Not enter `inMemoryIterator.hasNext` loop and `objectsWritten == 0`
    val dataSizes = {
      val spillBatchSize = conf.get(config.SHUFFLE_SPILL_BATCH_SIZE)
      Seq(spillBatchSize, spillBatchSize + 1, 0)
    }

    dataSizes.foreach { dataSize =>
      val dataBuffer = new PartitionedPairBuffer[Int, Int]
      (0 until dataSize.toInt).foreach(i => dataBuffer.insert(0, 0, i))
      val externalSorter = new TestExternalSorter[Int, Int, Int](taskContext)
      externalSorter.spill(dataBuffer)
    }

    // Verify recordsWritten same as data size
    assert(metricsCreated.length == dataSizes.length)
    metricsCreated.zip(dataSizes).foreach {
      case (metrics, dataSize) => assert(metrics.recordsWritten == dataSize)
    }

    // Verify bytesWritten same as file size
    assert(metricsCreated.length == filesCreated.length)
    filesCreated.foreach(file => assert(file.exists()))
    metricsCreated.zip(filesCreated).foreach {
      case (metrics, file) => assert(metrics.bytesWritten == file.length())
    }
  }
```

If the code in ExternalSorter keep original logic:
```scala
if (objectsWritten > 0) {
        flush()
        writer.close()
      } else {
        writer.revertPartialWritesAndClose()
      }
```

The above test suite will failed when `2. spillBatchSize + 1 -> objectsWritten > 0`, Expected file length is 40128 but actually it is 40134, the `suffix` did write to the end of the file. In other scenes, `suffix` was truncated.

When `objectsWritten > 0` also uses `writer.revertPartialWritesAndClose()`, all cases will pass.

**Comment by mridulm** [comment]
To add to my comment above - the issue is not whether there are additional suffix bytes written alone.
Rather, if there are suffix bytes written out which are not part of any of the segments (and close ends up being the trigger for footer write).
When `objectsWritten > 0`, then the footer goes in with the rest of the segment - but when `objectsWritten == 0`, there is no segment added for that footer.

**Comment by LuciferYang** [comment]
@mridulm I still feel a little confused, maybe my understanding is poor,  I try to understand what you mean:

1. If the data is written and the written records count is not an integer multiple of `serializerBatchSize`(objectsWritten > 0), it is correct that the footer is appended to the end of the file.

2. If the data is written and the written records count is an integer multiple of `serializerBatchSize`(The first possibility of objectsWritten == 0), it is correct that the footer is not appended to the end of the file.

3. If no writes at all, this is another possibility of objectsWritten == 0, it is correct that the footer is not appended to the end of the file.

Should I understand that?

**Comment by mridulm** [comment]
You are right.

Basic idea is, are the bytes captured in a segment which is tracked or not (which is done in `flush`: note, the `flush` method within `spillMemoryIteratorToDisk` - not `writer.flush`)

When `objectsWritten == 0`, there is no `flush` invoked - and so if there are any bytes written out during close, those wont be tracked in our offets.

**Comment by LuciferYang** [comment]
I check the code again, and a new discovery was synchronized to you @mridulm @Ngone51 

The status of `DiskBlockObjectWriter` is lazy initializing, the 'initialize()' method will only be called passively when the `write()` method is called for the first time and set `initialized` as `true`.

So in `No writes at all` scene, neither `close()` nor `revertPartialWritesAndClose()` will actually trigger the call of `closeResources()` because `initialized` is false and `revertPartialWritesAndClose()` will truncate an empty file.

**Comment by LuciferYang** [comment]
> objectsWritten == 0, there is no flush invoked

For the reasons I explained above, if `No writes at all` scenario is no longer considered, `objectsWritten == 0` means that 'inMemoryIterator.hasNext' return `false` after `flush()` is called when 'objectsWritten == serializerBatchSize' , so there is `flush()` invoked.

**Comment by LuciferYang** [comment]
On the other hand, I found `DiskBlockObjectWriter` use `objOut` to write kv,  `objOut` will `close()` after `commitAndGet()` and reassign in next call of `write()`, but `mcs(ManualCloseOutputStream  is defined by us)` can still be written before call `closeResources()`. 

So the potential extra content writing may occur every time `commitAndGet()` is called(by `objOut.close()`)? If so, it seems that potential additional data will be tracked by each `FileSegment`.

If the potential extra content writing occur in `closeResources()`,  it seems only possible to write some data to `mcs` manually before call `mcs.close()`, if the we write extra content manually, shouldn't we truncate it...

**Comment by mridulm** [comment]
I am fine with the discussion, but just to reiterate what I mentioned [above](https://github.com/apache/spark/pull/33556#issuecomment-889507340), I would prefer to really not change these codepaths unless there is an actual issue. It took a while to stabilize these, and they are fairly critical to get right.

**Comment by LuciferYang** [comment]
@mridulm ok ~ Maybe I can try this change in our internal version first and there will be at least one hundred thousand Spark applications every day to verify whether it has bad case.

**Comment by LuciferYang** [comment]
@mridulm @Ngone51 I'll close this pr first. Thank you for your review ~

**Pull request closed without merging**